docs/source/module_configuration/staticsite.rst

```diff
@@ -111,3 +111,149 @@ Most of these options are not required, but are listed here for reference::
               build_output: dist  # overrides default directory of top-level path setting
         regions:
           - us-west-2
+
+`Authorization@Edge`
+--------------------
+
+Authorization, the function of specifying access rights to resources, is often
+required to help protect restricted content in web applications. In a
+`blog post <https://aws.amazon.com/blogs/networking-and-content-delivery/authorizationedge-how-to-use-lambdaedge-and-json-web-tokens-to-enhance-web-application-security/>`_
+written by Alex Tomic and Cameron Worrell an in-depth solution is described
+on how to accomplish this with CloudFront, `Lambda@Edge` functions, and a
+Cognito User Pool. A similar solution has been integrated directly into
+Runway, via the StaticSite module, to allow easy and repeatable idempotent
+launching of private static sites.
+
+The following diagram depicts a high-level overview of this solution.
+
+.. image:: ../images/staticsite/auth_at_edge/flow_diagram.png
+
+Here is how the solution works:
+
+    1. The viewer’s web browser is redirected to Amazon Cognito custom UI page to sign up and authenticate.
+    2. After authentication, Cognito generates and cryptographically signs a JWT then responds with a redirect containing the JWT embedded in the URL.
+    3. The viewer’s web browser extracts JWT from the URL and makes a request to private content (private/* path), adding Authorization request header with JWT.
+    4. Amazon CloudFront routes the request to the nearest AWS edge location. The CloudFront distribution’s private behavior is configured to launch a `Lambda@Edge` function on ViewerRequest event.
+    5. `Lambda@Edge` decodes the JWT and checks if the user belongs to the correct Cognito User Pool. It also verifies the cryptographic signature using the public RSA key for Cognito User Pool. Crypto verification ensures that JWT was created by the trusted party.
+    6. After passing all of the verification steps, `Lambda@Edge` strips out the Authorization header and allows the request to pass through to designated origin for CloudFront. In this case, the origin is the private content Amazon S3 bucket.
+    7. After receiving response from the origin S3 bucket, CloudFront sends the response back to the browser. The browser displays the data from the returned response.
+
+An example of a Auth@Edge static site configuration is as follows. All listed options are required::
+
+    deployments:
+      - modules:
+        - name: sample-app
+          path: sample-app
+          type: static
+          environments:
+            dev:
+              namespace: sample-app-dev
+              staticsite_auth_at_edge: true
+              staticsite_aae_user_pool_arn: arn:aws:cognito-idp:us-east-1:240134083525:userpool/us-east-1_cjVgcUyWV
+        regions:
+          # NOTE: At the time of this writing Lambda@EDGE is only
+          # available in us-east-1. This solution will fail in
+          # other regions
+          - us-east-1
```

Comment:
instead of letting the deployment fail, can we add logic to `module/staticsite.py` that will log a warning when `staticsite_auth_at_edge and  region != 'us-east-1` and skip it? it would be more user friendly.

Or, if we do want to just fail out when someone tries to do it. throw a custom error.

I wouldn't be too concerned with _what if they extend support to other regions_. there are plenty of features that have been around for years that only operate in us-east-1.

Reply:
I think AWS already handles this for us. I'll have to double check but last I remember it gives a warning stating that you can't upload Lamdba@Edge functions outside of us-east-1. I just have the added documentation there so they don't run into the error message like I did.

If we skip instead of failing outright then we'd have to have it launch as a normal static site, I don't think that's "bad" persay but I also don't think it's what a user would want. I personally would expect it to fail on me instead of loading up something I didn't ask for.

No concern around it going outside us-east-1, based on the design of using the closest location I don't think they'd ever need a reason to put it in another region.

Reply:
>  AWS already handles this for us ... so they don't run into the error message like I did

Thats what i would like to avoid, the user getting an error from AWS they have to figure out. If we can anticipate it on our end we can provide something more informative before we try to hit AWS.

> If we skip instead of failing outright then we'd have to have it launch as a normal static site

I would skip the module entirely, just `return` right after we identify this case and let runway move onto the next module (with a warning logged for why it was skipped). I wouldn't want to deploy a public site in place since that is asking for data leaks.

Reply:
> I would skip the module entirely, just return right after we identify this case and let runway move onto the next module 

Couldn't this cause problems with cross-module dependencies? Not that I assume there will be a lot of them from a static site, but in the rare occasion there is wouldn't you want it to fail?

Reply:
Spoke with @ITProKyle and @troy away from thread. We agreed to kill the process instead of moving to the next. Implementing shortly